### PR TITLE
feat: support namespace analysis

### PR DESCRIPTION
**Pull request by JustAGabriel**
The "XML Validator" tool is now able to detect inconsistencies in XML namespace references for both the XML and the XSD data.

This capability will prevent "always true" validation cases potentially caused by validating XML data without the correct namespace scopes.

Refs: #604

## Pull request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] UI change (please include screenshot!)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Internationalization and localization
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: #604

## What is the new behavior?

The tool is now able to detect unreflected namespace reference and issue an according warning.
Example: The XML data defines a "xmlns" namespace not defined in the XSD => result of the data validation will always be 'valid', a warning is ejected.

## Quality check

Before creating this PR, have you:

- [x] Followed the code style guideline as described in [CONTRIBUTING.md](https://github.com/veler/DevToys/blob/main/CONTRIBUTING.md)
- [x] Verified that the change work in Release build configuration
- [x] Checked all unit tests pass

### CODE REVIEW

File: src/dev/impl/DevToys/Helpers/XmlHelper.cs [hunk 1]
```diff
@@ -1,10 +1,12 @@
 ﻿#nullable enable
 
 using System;
-using System.IO;
+using System.Collections.Generic;
```
Comment (veler) [inline src/dev/impl/DevToys/Helpers/XmlHelper.cs:4]: These usings seem unnecessary.

File: src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
+﻿using System;
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs:1]: Please add `#nullable enable` at the top of the file.

File: src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            missingTargetNamespaceUri = xsdParsingResult.TargetNamespace;
+            return true;
+        }
+
+        private static IEnumerable<XmlNamespace> GetNamespacesFromXsd(XsdParsingResult xsdParsingResult)
+        {
+            static bool IsDefaultXsdNamespace(XmlNamespace xmlNamespace) => xmlNamespace.Prefix.StartsWith("xs", StringComparison.InvariantCultureIgnoreCase);
+
+            // return namespaces without default XSD namespace (xmlns:xs="http://www.w3.org/2001/XMLSchema")
+            return xsdParsingResult.Namespaces.Where(xns => !IsDefaultXsdNamespace(xns)).ToList();
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs:62]: `.ToList();` does a copy of the result in memory, which can be expensive. Looking at the code, it doesn't look like a copy is needed.

File: src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        /// </summary>
+        /// <param name="xsdParsingResult">XSD parsing result</param>
+        /// <param name="xmlParsingResult">XML parsing result</param>
+        /// <returns>The missing namespaces</returns>
+        public static List<XmlNamespace> GetMissingNamespacesInXml(XsdParsingResult xsdParsingResult, XmlParsingResult xmlParsingResult)
+        {
+            IEnumerable<XmlNamespace> clearedXdsNamespaces = GetNamespacesFromXsd(xsdParsingResult);
+            IEnumerable<XmlNamespace> xmlNamespaces = xmlParsingResult.Namespaces;
+
+            return clearedXdsNamespaces.Where(ns => xmlNamespaces.Contains(ns) == false).ToList();
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs:41]: Same thing here

File: src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            IEnumerable<XmlNamespace> clearedXdsNamespaces = GetNamespacesFromXsd(xsdParsingResult);
+            IEnumerable<XmlNamespace> xmlNamespaces = xmlParsingResult.Namespaces;
+
+            // filter targetNamespace
+            if (xsdParsingResult.TargetNamespace is not null)
+            {
+                xmlNamespaces = xmlNamespaces.Where(xns => !string.Equals(xns.Uri, xsdParsingResult.TargetNamespace, StringComparison.InvariantCultureIgnoreCase)).ToList();
+            }
+
+            return xmlNamespaces.Where(ns => clearedXdsNamespaces.Contains(ns) == false).ToList();
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs:27]: And here

File: src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/Parsing/XsdParser.cs [hunk 6]
```diff
@@ -12,12 +12,15 @@ public XsdParser(string source) : base(source)
...
 
-        internal override XsdParsingResult Parse(string xsdContent)
+        protected override XsdParsingResult ParsingOperation(string xsdContent)
         {
             XmlSchema xmlSchema;
+            string? targetNamespace;
             try
             {
-                xmlSchema = XmlSchema.Read(new XmlTextReader(new StringReader(xsdContent)), ValidationErrorCallBack);
+                StringReader reader = new(xsdContent);
```
Comment (veler) [inline src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/Parsing/XsdParser.cs:21]: Make sure to `Dispose()` the reader.

**Review by veler — COMMENTED** [review]
Hi,
Thank you for this PR! It looks great overall 😁
A few comments here and there to address, and we can merge it 😊
Thanks a lot again!

File: src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs (hunk 2 above)
Comment (JustAGabriel) [reply src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs]: done.

File: src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs (hunk 3 above)
Comment (JustAGabriel) [reply src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs]: Oh, great catch! 
Refactored the sections accordingly.

File: src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/Parsing/XsdParser.cs (hunk 6 above)
Comment (JustAGabriel) [reply src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/Parsing/XsdParser.cs]: added a `using` statement here - does this work for you?

File: src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs (hunk 4 above)
Comment (JustAGabriel) [reply src/dev/impl/DevToys/ViewModels/Tools/Text/XmlValidator/NamespaceHelper.cs]: done.

**Comment by JustAGabriel** [comment]
Hi Etienne!
Thanks so much for the fast review!
I addressed all of your remarks - hope you find everything as expected.
BR!

**Review by veler — APPROVED** [review]
Awesome! Thank you so much for addressing this feedback :D 
I will merge the changes. Thank you for your contribution !
An update (1.0.10.0) is schedule to be released tomorrow. It won't include your changes though. They will go into the next update (likely 1.0.11.0)

**Pull request merged**